docs/lang/articles/advanced/odop.md

```diff
@@ -15,47 +15,206 @@ object-oriented programming (OOP).
+class TiArray:
+    def __init__(self, n):
+        self.x = ti.field(dtype=ti.i32, shape=n)
+
+    @ti.kernel
+    def inc(self):
+        for i in self.x:
+            self.x[i] += 1.0
+
+a = TiArray(32)
+a.inc()
+```
+
+After new feature **Dynamic SNode** released in `v0.8.0`, you could define **Taichi Fields** not only in `__init__` but also **any place** of Python-scope functions, like:
+
+```python
+import taichi as ti
+
+ti.init(arch=ti.cuda)
+
+@ti.data_oriented
+class MyClass:
+    @ti.kernel
+    def func(self, temp: ti.template()):
+        for i in ti.grouped(temp):
+            temp[i] += 1
+
+    def call_func(self):
+        self.func(self.temp)
+
+    def allocate_temp(self, n):
+        self.temp = ti.field(dtype = ti.i32, shape=n)
+
+
+
+a = MyClass()
+# a.call_func()
+# temp has not been allocated now
+a.allocate_temp(4)
+a.call_func()
+a.call_func()
+print(a.temp)
+a.allocate_temp(8)
+a.call_func()
+print(a.temp)
+```
+
+Another example of automatically recycle memory:
+```python
+import taichi as ti
+
+ti.init(arch=ti.cuda)
+
+@ti.data_oriented
+class Calc:
+    def __init__(self):
+        self.x = ti.field(dtype=ti.f32, shape=16)
+        self.y = ti.field(dtype=ti.f32, shape=4)
+
+    @ti.kernel
+    def func(self, temp: ti.template()):
+        for i in range(8):
+            temp[i] = self.x[i * 2] + self.x[i * 2 + 1]
+
+        for i in range(4):
+            self.y[i] = max(temp[i * 2], temp[i * 2 + 1])
+
+    def call_func(self):
+        fb = ti.FieldsBuilder()
+        temp = ti.field(dtype=ti.f32)
+        fb.dense(ti.i, 8).place(temp)
+        tree = fb.finalize()
+        self.func(temp)
+        tree.destroy()
+
+
+
+a = Calc()
+for i in range(16):
+    a.x[i] = i
+a.call_func()
+print(a.y)
+```
+
+
+## Some workaround about Python
+
+### Inherit of data-oriented class
+
+The *data-oriented* property will be automatically carried beyond the Python class inheriting. This means the **Taichi Kernel** could be called while any of the ancestor classes are decorated by the `@ti.data_oriented` decorator.
+
+An example:
+```python
+import taichi as ti
+
+ti.init(arch=ti.cuda)
+
+class BaseClass:
+    def __init__(self):
+        self.n = 10
+        self.num = ti.field(dtype=ti.i32, shape=(self.n, ))
+
+    @ti.kernel
+    def count(self) -> ti.i32:
+        ret = 0
+        for i in range(self.n):
+            ret += self.num[i]
+        return ret
+
+    @ti.kernel
+    def add(self, d: ti.i32):
+        for i in range(self.n):
+            self.num[i] += d
+
+
+@ti.data_oriented
+class DataOrientedClass(BaseClass):
+    pass
+
+class DeviatedClass(DataOrientedClass):
+    @ti.kernel
+    def sub(self, d: ti.i32):
+        for i in range(self.n):
+            self.num[i] -= d
+
+
+a = DeviatedClass()
+a.add(1)
+a.sub(1)
+print(a.count()) # Output: 0
+
+
+b = DataOrientedClass()
+b.add(2)
+print(b.count()) # Output: 1
+
+c = BaseClass()
+# c.add(3)
+# print(c.count())
+# These two lines would cause error...
```

Comment:
Grammar-wise you may want to change this line to 
```
# The two lines above will trigger a (what kind of) error.
```

Please specify the type of error for more clarity.